### PR TITLE
Feature/235 catch jwt logic

### PR DESCRIPTION
**Pull request by Rebal67**

### CODE REVIEW

File: src/global.d.ts [hunk 1]
```diff
@@ -0,0 +1,7 @@
+import { JSDOM } from 'jsdom';
+
+declare global {
```
Comment (theblockstalk) [inline src/global.d.ts:3]: I dont think this should be in SRC as its about the test

File: src/app.ts [hunk 2]
```diff
@@ -107,9 +112,31 @@ export default class App {
         if (redirect) {
             // const settings = await getSettings();
             // TODO update settings to redirect to the tonomy id website
-            window.location = `https://id.tonomy.com/login?jwt=${token}`;
+            window.location = `https://localhost:3000/login?jwt=${token}`;
```
Comment (theblockstalk) [inline src/app.ts:115]: this should come from settings

File: src/app.ts [hunk 3]
```diff
@@ -107,9 +112,31 @@ export default class App {
...
+
+    static async onRedirectLogin(): Promise<JWTLoginPayload> {
+        const urlParams = new URLSearchParams(window.location.search);
+        const jwt = urlParams.get('jwt');
+        if (!jwt) throwError('No JWT found in URL');
+        const verified = await this.verifyLoginJWT(jwt);
+        const payload = verified.payload as JWTLoginPayload;
+
+        if (payload.origin !== document.referrer)
+            throwError(`JWT origin: ${payload.origin} does not match referrer: ${document.referrer}`);
```
Comment (theblockstalk) [inline src/app.ts:129]: can you add a new SdkErrors enum value and add the code here

File: src/app.ts [hunk 4]
```diff
@@ -107,9 +112,31 @@ export default class App {
...
+        return payload;
+    }
+
+    static async verifyLoginJWT(jwt: string): Promise<JWTVerified> {
+        const resolver: any = {
+            resolve,
+        };
+        const verify = await verifyJWT(jwt, { resolver });
```
Comment (theblockstalk) [inline src/app.ts:137]: combine this into one line. `return await verifyJWT...`

File: test/app.test.ts [hunk 5]
```diff
@@ -1,7 +1,10 @@
+/**
+ * @jest-environment-options {"url": "http://localhost:3000/login?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJFUzI1NkstUiJ9.eyJpYXQiOjE2Njk0NTkzNzgsIm51bWJlciI6ImE5MmRkNWQ3MmQ5NDAyMDNkNjRjZjJlYzRmY2Q3N2NjYzQyZjNkZWRlNDBiNjFiNmJiNDM2YzgyOGU0OTQ4MjUiLCJvcmlnaW4iOiJsb2NhbGhvc3QiLCJwdWJrZXkiOiJQVUJfSzFfODcyMmVnaWluYWE0bjIyZTZIQmNBQzlnaHpaOW5Qa2pqQzNwdFAxOWtoMzd4NzlwQ1YiLCJpc3MiOiJkaWQ6andrOmV5SmpjbllpT2lKelpXTndNalUyYXpFaUxDSnJkSGtpT2lKRlF5SXNJbmdpT2lKd05HeHBlR1YxUVZsSVVrVkVWR1JoYTJsR2NtTkliaXRSVkVVNE5DOVhRM0p4Unk5UmNqQkNkVmx6UFNJc0lua2lPaUoyVldOdVlTdDNhbGR2UWpoeWVGaElUVkJCV1V4UWN6RjRZbE14ZEZnMmMzSm1WRXBOVDBrclpHZEZQU0lzSW10cFpDSTZJbEJWUWw5TE1WODROekl5WldkcGFXNWhZVFJ1TWpKbE5raENZMEZET1dkb2VsbzVibEJyYW1wRE0zQjBVREU1YTJnek4zZzNPWEJEVmlKOSJ9.MW1J8SjTwSufE_n5fhPhhA9rjSY4MEbyD2WM-KswMNThGjK8tWOoRByrEZR8r_kl0mo7xe8zQDRG8NEwPKyiZwE
```
Comment (theblockstalk) [inline test/app.test.ts:2]: delete this comment? this string is duplicated below

File: test/app.test.ts [hunk 6]
```diff
@@ -13,4 +16,23 @@ describe('logging in', () => {
         const jwt = await App.onPressLogin(window);
         expect(jwt).toBeDefined();
     });
+
+    it('checks login url', async () => {
+        const jwt =
+            'eyJ0eXAiOiJKV1QiLCJhbGciOiJFUzI1NkstUiJ9.eyJpYXQiOjE2Njk0NTkzNzgsIm51bWJlciI6ImE5MmRkNWQ3MmQ5NDAyMDNkNjRjZjJlYzRmY2Q3N2NjYzQyZjNkZWRlNDBiNjFiNmJiNDM2YzgyOGU0OTQ4MjUiLCJvcmlnaW4iOiJsb2NhbGhvc3QiLCJwdWJrZXkiOiJQVUJfSzFfODcyMmVnaWluYWE0bjIyZTZIQmNBQzlnaHpaOW5Qa2pqQzNwdFAxOWtoMzd4NzlwQ1YiLCJpc3MiOiJkaWQ6andrOmV5SmpjbllpT2lKelpXTndNalUyYXpFaUxDSnJkSGtpT2lKRlF5SXNJbmdpT2lKd05HeHBlR1YxUVZsSVVrVkVWR1JoYTJsR2NtTkliaXRSVkVVNE5DOVhRM0p4Unk5UmNqQkNkVmx6UFNJc0lua2lPaUoyVldOdVlTdDNhbGR2UWpoeWVGaElUVkJCV1V4UWN6RjRZbE14ZEZnMmMzSm1WRXBOVDBrclpHZEZQU0lzSW10cFpDSTZJbEJWUWw5TE1WODROekl5WldkcGFXNWhZVFJ1TWpKbE5raENZMEZET1dkb2VsbzVibEJyYW1wRE0zQjBVREU1YTJnek4zZzNPWEJEVmlKOSJ9.MW1J8SjTwSufE_n5fhPhhA9rjSY4MEbyD2WM-KswMNThGjK8tWOoRByrEZR8r_kl0mo7xe8zQDRG8NEwPKyiZwE';
+        const url = 'http://localhost/login?jwt=' + jwt;
+
+        // eslint-disable-next-line prettier/prettier
```
Comment (theblockstalk) [inline test/app.test.ts:25]: delete L25 and L27 comments

File: src/global.d.ts (hunk 1 above)
Comment (Rebal67) [reply src/global.d.ts:3]: without this jsdom won't resolve iif put inside test folder it wont do anything

File: src/app.ts (hunk 2 above)
Comment (Rebal67) [reply src/app.ts:115]: settings isn't global it cant be used here we have a task for this

**Pull request merged**